server/src/main/java/org/opensearch/search/query/QueryCollectorContextSpecRegistry.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.query;
+
+import org.opensearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * Registry class to load all collector context spec factories during cluster bootstrapping
+ */
+public class QueryCollectorContextSpecRegistry {
+    private static final List<QueryCollectorContextSpecFactory> registry = new CopyOnWriteArrayList<>();
+
+    /**
+     * Get all collector context spec factories
+     * @return list of collector context spec factories
+     */
+    public static List<QueryCollectorContextSpecFactory> getCollectorContextSpecFactories() {
+        return registry;
+    }
+
+    /**
+     * Register factory
+     * @param factory collector context spec factory defined in plugin
+     */
+    public static void registerFactory(QueryCollectorContextSpecFactory factory) {
+        registry.add(factory);
+    }
+
+    /**
+     * Get collector context spec
+     * @param searchContext search context
+     * @param queryCollectorArguments query collector arguments
+     * @return collector context spec
+     * @throws IOException
+     */
+    public static QueryCollectorContextSpec getQueryCollectorContextSpec(
+        final SearchContext searchContext,
+        final QueryCollectorArguments queryCollectorArguments
+    ) throws IOException {
+        Iterator<QueryCollectorContextSpecFactory> iterator = registry.iterator();
+        while (iterator.hasNext()) {
+            QueryCollectorContextSpecFactory factory = iterator.next();
+            Optional<QueryCollectorContextSpec> spec = factory.createQueryCollectorContextSpec(searchContext, queryCollectorArguments);
+            if (spec != null && spec.isPresent()) {
```

Comment:
Super nit, the `null` check is redundant (this is why we use `Optional`)

Reply:
I would suggest lets atleast keep isPresent check.

Reply:
And return spec.get().

Reply:
`spec.isPresent()` is totally fine, I am calling `null` only, thank you